docs/auth.md

```diff
@@ -0,0 +1,96 @@
+## User Authentication in Soul
+
+Soul incorporates a robust user authentication system that handles user accounts, groups, permissions, and cookie-based user sessions. This section provides an overview of how the default implementation works.
+
+### Overview
+
+The Soul authentication system handles both authentication and authorization. Briefly, authentication verifies a user is who they claim to be, and authorization determines what an authenticated user is allowed to do. Here, the term "authentication" is used to refer to both tasks.
+
+The auth system consists of:
+
+- Users
+- Permissions: Binary (yes/no) flags designating whether a user may perform a certain task.
+- Default Permissions
+- Roles: A generic way of applying labels and permissions to more than one user.
+- A password hashing system
+- APIs for logging in users or restricting content
+
+The authentication system in Soul aims to be very generic and doesn't provide some features commonly found in web authentication systems, such as:
+
```

Comment:
I've always imagined Soul as a flexible and modular framework, and the introduction of API extensions is a step in that direction. I've also been thinking about adding Middleware extensions, which would empower Soul users to create their own Express middlewares and customize their experience.

Another idea I've been considering is gradually integrating these features into Soul itself, while still giving users the ability to configure and enable/disable them according to their preferences. But for now, I think releasing the first version of authentication/authorization as described in this documentation, even in its minimalist form, would be a major milestone.

Reply:
> But for now, I think releasing the first version of authentication/authorization as described in this documentation, even in its minimalist form, would be a major milestone.

I fully understand. But, my users have a set of constraints for password-checking, which we have implemented in JS (as an _extension).  I would wish/need to include this in the auth implementation, and this would be easier if the new auth allowed me to _plug-in_ a method providing the additional functionality.

Reply:
That's totally fine! If it's just about checking the password, we can easily include it in the new authentication design. And if there's more to it, we can figure out how to handle it together.

By the way, can you share your extensions with me?

Reply:
> By the way, can you share your extensions with me?

Sure, extensions [here](https://github.com/DeepBlueCLtd/RCO/tree/main/_devExtensions)

Reply:
I believe that you can achieve some of these once we implement the `*Middleware* extensions`.

Reply:
I _guess_ the `Middleware` is a future feature.  I'd like to include proposed password checking  in this phase, if we can.  Could there be the option to provide an optional `allow` method when a password is updated?

Hmm, my app currently uses `vanilla` Soul. So, we can't specify such a method by modifying Soul-cli.  We just have the declared `extensions`.  So, I _guess_ we could either provide the name of an extension in the startup command, or declare an extension using a `special` name: `_allowPassword`?

Hmm, no. This won't be an _extension_ that receives a req and returns a response.  It would just be a method that receives a user id and the proposed password, and returns true/false. (or either a null (success) or an error message (failure) ).

Aah, or maybe it could be an _extension_.  If an `_updatePassword` extension is provided, let that handle the password update, else handle it with a routine update.  I can provide `_updatePassword` extension, and it can do new password validation and update the value if successful.

Reply:
We can try to implement a basic version of middleware extensions.